nf_core/pipelines/downloads/singularity.py

```diff
@@ -0,0 +1,573 @@
+            if shutil.which("singularity"):
+                singularity_command = [
+                    "singularity",
+                    "pull",
+                    "--name",
+                    output_path_tmp.name,
+                    address,
+                ]
+            elif shutil.which("apptainer"):
+                singularity_command = ["apptainer", "pull", "--name", output_path_tmp.name, address]
+            else:
+                raise OSError("Singularity/Apptainer is needed to pull images, but it is not installed or not in $PATH")
+            log.debug(f"Building singularity image: {address}")
+            log.debug(f"Singularity command: {' '.join(singularity_command)}")
+
+            # Progress bar to show that something is happening
+            task = self.progress.add_task(
+                container,
+                start=False,
+                total=False,
+                progress_type="singularity_pull",
+                current_log="",
+            )
+
+            # Run the singularity pull command
+            with subprocess.Popen(
+                singularity_command,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                universal_newlines=True,
+                bufsize=1,
+            ) as proc:
+                lines = []
+                if proc.stdout is not None:
+                    for line in proc.stdout:
+                        lines.append(line)
+                        self.progress.update(task, current_log=line.strip())
+
+            if lines:
+                # something went wrong with the container retrieval
+                if any("FATAL: " in line for line in lines):
+                    self.progress.remove_task(task)
+                    raise ContainerError(
+                        container=container,
+                        registry=library,
+                        address=address,
+                        absolute_URI=absolute_URI,
+                        out_path=output_path,
+                        singularity_command=singularity_command,
+                        error_msg=lines,
+                    )
+
+        self.symlink_registries(output_path)
+        self.progress.remove_task(task)
+
+    def copy_image(self, container: str, src_path: str, dest_path: str) -> None:
+        """Copy Singularity image from one directory to another."""
+        log.debug(f"Copying {container} from '{os.path.basename(src_path)}' to '{os.path.basename(dest_path)}'")
+
+        with intermediate_file(dest_path) as dest_path_tmp:
+            shutil.copyfile(src_path, dest_path_tmp.name)
+
+        # Create symlinks to ensure that the images are found even with different registries being used.
+        self.symlink_registries(dest_path)
+
+    def fetch_containers(
+        self,
+        containers: Collection[str],
+        output_dir: str,
+        exclude_list: Container[str],
```

Comment:
That is what I called nuanced typing 🫡. It is true, the `exclude_list` does not need to be iterable.